### PR TITLE
Refactor jit, expand documentation

### PR DESCRIPTION
**Pull request by tcapelle**
- Some initial refactor, and documentation
- Add `cpu` option to `to_jit` func
- Add `export_dls` to pickel the dataloader
- `new_export` as POC of APi.

### CODE REVIEW

File: fastexport/jit.py [hunk 1]
```diff
@@ -8,24 +8,20 @@
...
-    return self
-
 # Cell
 mk_class('JitMode', **{o:o.lower() for o in ['Trace','Script']},
          doc="All possible export modes as attributes to get tab-completion and typo-proofing")
 
 # Cell
 @patch
-def to_jit(self:Learner, fname='export.pt', mode=JitMode.Trace):
+def to_jit(self:Learner, fname='torchscript_model.ts', mode=JitMode.Trace, cpu=False):
```
Comment (muellerzr) [inline fastexport/jit.py]: Let's get rid of `cpu=False`. `learn.export` will handle the device, so switch that to `device` and we can default it to `cpu`

File: fastexport/jit.py [hunk 2]
```diff
@@ -8,24 +8,20 @@
...
     "Exports `learn.model` using `jit` with `mode` to `fname`"
     inp = self.dls.one_batch()[:self.dls.n_inp]
     if not isinstance(inp, tuple): inp = (inp,)
     self.model.eval()
     self.model.to(inp[0].device)
```
Comment (muellerzr) [inline fastexport/jit.py]: Let's simplify this logic given the above assumption:
```python
self.model.to(device)
inp = to_device(inp, device)
```

File: fastexport/jit.py [hunk 3]
```diff
@@ -8,24 +8,20 @@
...
+    if cpu:
+        self.model.cpu()
+        inp = to_device(inp, 'cpu')
     traced_model = getattr(torch.jit, mode)(self.model, inp)
-    torch.jit.save(traced_model, fname)
\ No newline at end of file
+    torch.jit.save(traced_model, str(fname))
```
Comment (muellerzr) [inline fastexport/jit.py]: Are we sure we need this cast to a string? fastai's export works with a pathlib by doing:
```python
learn.path/fname
```
Which we also should mimic

Comment (muellerzr) [reply fastexport/jit.py]: (Which is just `torch.save`)

File: 00_jit.ipynb [hunk 4]
```diff
@@ -53,180 +63,56 @@
...
    "source": [
-    "#slow\n",
-    "set_seed(99, True)\n",
-    "path = untar_data(URLs.PETS)/'images'\n",
-    "dls = ImageDataLoaders.from_name_func(\n",
-    "    path, get_image_files(path), valid_pct=0.2,\n",
-    "    label_func=lambda x: x[0].isupper(), item_tfms=Resize(224))"
+    "from torch doc: \n",
+    "\n",
+    "**In many cases either tracing or scripting is an easier approach for converting a model to TorchScript. Tracing and scripting can be composed to suit the particular requirements of a part of a model\n",
```
Comment (muellerzr) [inline 00_jit.ipynb]: Use a block quote rather than bold

`>`

File: 00_jit.ipynb [hunk 5]
```diff
@@ -53,180 +63,56 @@
...
-   "execution_count": null,
+   "cell_type": "markdown",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "#slow\n",
-    "from fastai.vision.all import *"
+    "There are two possible scenarios with `jit`: `trace` and `script`. \n",
+    "\n",
+    "**I don't think there is an ideal** Ideally you should use `torch.jit.trace`, however if there is dynamic behavior, `torch.jit.script` should be utilized instead. As a result `trace` is tried by default."
```
Comment (muellerzr) [inline 00_jit.ipynb]: Phrasing, You say "I don't think there's an ideal" and then say "Ideally".

File: 00_jit.ipynb [hunk 6]
```diff
@@ -389,6 +284,68 @@
     "> As these are just the models, raw probabilities are returned. You still need to perform a soft or argmax"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You could also save a cpu version of the model:"
```
Comment (muellerzr) [inline 00_jit.ipynb]: Don't add cpu version example, specify the flag explicitly in the first example to be efficient.

File: 00_jit.ipynb [hunk 7]
```diff
@@ -411,12 +368,33 @@
...
+    "def label_func(x): return x[0].isupper()\n",
     "dls = ImageDataLoaders.from_name_func(\n",
-    "    path, get_image_files(path), valid_pct=0.2,\n",
-    "    label_func=lambda x: x[0].isupper(), item_tfms=Resize(224))\n",
+    "    '.', get_image_files(path), valid_pct=0.2,\n",
+    "    label_func=label_func, item_tfms=Resize(224))\n",
     "learn = cnn_learner(dls, resnet18)"
    ]
   },
+  {
```
Comment (muellerzr) [inline 00_jit.ipynb]: We don't need this here.

File: 00_jit.ipynb [hunk 8]
```diff
@@ -480,12 +458,193 @@
...
     "x,_ = dls.one_batch()\n",
     "\n",
     "probs = loaded_model(x); probs[:3]"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exporting the dataloader and transforms\n",
```
Comment (muellerzr) [inline 00_jit.ipynb]: Move exporting the DataLoaders to its own notebook and its own PR, there's much to adjust there and it's not jit specific. Keep each PR small and incremental - Jeremy

File: 00_jit.ipynb [hunk 9]
```diff
@@ -480,12 +458,193 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@patch\n",
+    "@delegates(Learner.to_jit)\n",
```
Comment (muellerzr) [inline 00_jit.ipynb]: Move export to its own notebook, as this is not jit specific. And it's own PR

**Review by muellerzr — CHANGES_REQUESTED** [review]
I made some comments I'd like fixed, and we really should separate these out rather than one very big PR. Harder to track little mistakes and to review.

**Comment by muellerzr** [comment]
Also ensure you bring back the TensorBase patch, we need it until the next release of fastai for tests to pass on the CI

File: 00_jit.ipynb (hunk 5 above)
Comment (tcapelle) [reply 00_jit.ipynb]: lol this was just to comment that the phrase :Ideally you should use" is not correct

**Pull request merged**